### ISSUE TITLE
Fix warnings triggered by Pytest in Python unit tests

### ISSUE THREAD

**Issue by AdriiiPRodri** [wazuh/wazuh#13267]
In this issue we will investigate and try to solve the warnings that unit and integration tests throw. This is an example of warnings of the framework's unittests.

```
var/ossec/framework/python/lib/python3.9/site-packages/jsonschema/compat.py:6
  /var/ossec/framework/python/lib/python3.9/site-packages/jsonschema/compat.py:6: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.10 it will stop working
    from collections import MutableMapping, Sequence  # noqa

var/ossec/framework/python/lib/python3.9/site-packages/connexion/apis/aiohttp_api.py:48
  /var/ossec/framework/python/lib/python3.9/site-packages/connexion/apis/aiohttp_api.py:48: DeprecationWarning: "@coroutine" decorator is deprecated since Python 3.8, use "async def" instead
    def problems_middleware(request, handler):

var/ossec/framework/python/lib/python3.9/site-packages/connexion/apis/aiohttp_api.py:169
  /var/ossec/framework/python/lib/python3.9/site-packages/connexion/apis/aiohttp_api.py:169: DeprecationWarning: "@coroutine" decorator is deprecated since Python 3.8, use "async def" instead
    def _get_openapi_json(self, request):

var/ossec/framework/python/lib/python3.9/site-packages/connexion/apis/aiohttp_api.py:177
  /var/ossec/framework/python/lib/python3.9/site-packages/connexion/apis/aiohttp_api.py:177: DeprecationWarning: "@coroutine" decorator is deprecated since Python 3.8, use "async def" instead
    def _get_openapi_yaml(self, request):

var/ossec/framework/python/lib/python3.9/site-packages/connexion/apis/aiohttp_api.py:236
  /var/ossec/framework/python/lib/python3.9/site-packages/connexion/apis/aiohttp_api.py:236: DeprecationWarning: "@coroutine" decorator is deprecated since Python 3.8, use "async def" instead
    def _get_swagger_ui_home(self, req):

var/ossec/framework/python/lib/python3.9/site-packages/connexion/apis/aiohttp_api.py:246
  /var/ossec/framework/python/lib/python3.9/site-packages/connexion/apis/aiohttp_api.py:246: DeprecationWarning: "@coroutine" decorator is deprecated since Python 3.8, use "async def" instead
    def _get_swagger_ui_config(self, req):

var/ossec/framework/python/lib/python3.9/site-packages/connexion/apis/aiohttp_api.py:295
  /var/ossec/framework/python/lib/python3.9/site-packages/connexion/apis/aiohttp_api.py:295: DeprecationWarning: "@coroutine" decorator is deprecated since Python 3.8, use "async def" instead
    def get_request(cls, req):

var/ossec/framework/python/lib/python3.9/site-packages/connexion/apis/aiohttp_api.py:324
  /var/ossec/framework/python/lib/python3.9/site-packages/connexion/apis/aiohttp_api.py:324: DeprecationWarning: "@coroutine" decorator is deprecated since Python 3.8, use "async def" instead
    def get_response(cls, response, mimetype=None, request=None):

var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.4.0-py3.9.egg/wazuh/tests/test_cluster.py:63
  /var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.4.0-py3.9.egg/wazuh/tests/test_cluster.py:63: PytestUnknownMarkWarning: Unknown pytest.mark.asyncio - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.asyncio

var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.4.0-py3.9.egg/wazuh/tests/test_cluster.py:79
  /var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.4.0-py3.9.egg/wazuh/tests/test_cluster.py:79: PytestUnknownMarkWarning: Unknown pytest.mark.asyncio - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.asyncio
```

**Note: This should be fixed once we upgrade to version 3.10 of Python.**

**Comment by GGP1** [comment]
## Update

Started debugging warnings and errors in an environment with the following versions:

```
Python 3.10.6
pip 23.0.1
Master branch
```

Found 31 warnings and 13 errors in total, only 3 of each type are unique.

### Warnings

- [x] DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives
- [ ] PytestUnknownMarkWarning: Unknown pytest.mark.asyncio - is this a typo?  You can register custom marks to avoid this warning
- [ ] DeprecationWarning: "@coroutine" decorator is deprecated since Python 3.8, use "async def" instead

### Errors

- [x] ModuleNotFoundError: No module named 'freezegun'
- [ ] ImportError: from .loop import Loop as __BaseLoop  # NOQA
- [ ] ImportError: cannot import name 'Callable' from 'collections'

**Comment by GGP1** [comment]
## Status update

Total warnings: 7
Warnings from dependencies: 6
Warnings from Wazuh's source code: 1
Total errors: 0

| Module | Warnings |
| ----------- | -------------- |
| Framework | 2 |
| API | 3 |
| Cluster | 1 |
| Wodles | 1 |

### Warnings

 - [ ] [framework|API|cluster] Runtime: coroutine 'x' was never awaited. 

      From **unittest** and **_pytest** packages.

- [ ] [framework] Future: Possible nested set at position 2.

     Regular expression format warning at `wazuh/framework/wazuh/rbac/auth_context.py:225`.
       
- [ ] [API] PytestUnknownMark: Unknown pytest.mark._standalone/cluster_. From the **_pytest** package.
       
- [ ] [wodles] Deprecation: The distutils package is deprecated and slated for removal in Python 3.12. From the package **google/pubsub_v1/services/publisher**.

**Comment by GGP1** [comment]
## Update

The status has been moved to "On hold" until we define how we will continue with this issue and the ones related to it, we expect to have a state update on Monday.

**Comment by GGP1** [comment]
## Update

The issue is being delayed due to the epic issue assignees having higher priority tasks. We will resume the development once we are available and can coordinate the lasts steps to take in order to migrate the Python version to 3.10.

Closed the related issue https://github.com/wazuh/wazuh/issues/12159 since wodles' warnings are already solved in this one.

**Comment by mhamra** [comment]
## Update

I was analyzing the warning "RuntimeWarning: coroutine 'x' was never awaited. From unittest and _pytest packages." raised by test framework/wazuh/core/cluster/tests/test_client.py and found what is the reason for this warning.

The test "test_ac_connection_made" patches async.gather function and sets its return value as a mocked task. The test calls the function AbstractClient.connection_made:
```
    def connection_made(self, transport):
        """Define process of connecting to the server.

        Parameters
        ----------
        transport : asyncio.Transport
            Socket to write data on.
        """
        self.transport = transport
        future = asyncio.gather(self.log_exceptions(self.send_request(command=b'hello', data=self.client_data)))
        future.add_done_callback(self.connection_result)
```

asyncio.gather launches a new task by calling log_exceptions and send_requests. asyncio.gather return value is a Future to be awaited. In connection_made this is done by calling add_done_callback. Because the function asyncio.gather was patched, its return value is a mocked Future, but the mocked future does not set a callback to another function, nor awaits the return value of the Future.

**Comment by mhamra** [comment]
## UPDATE

- Checking framework/requirements.txt, there is a dependency problem that is solved by changing the request package from 2.25.1 to 2.26.0 
- Running unit tests with Python 3.10.11, two tests fail. both failures were fixed.

```
- wazuh/core/cluster/tests/test_master.py::test_master_handler_hello_ok
- wazuh/core/tests/test_wdb.py::test_async_send_ko
```

- fixed tests: 
```
- test_client.py, 
- test_local_client.py, 
- test_common.py, 
- test_local_server.py, 
- test_master.py
```
- Work in progress: test_server.py

**Comment by GGP1** [comment]
## Update

- Created an issue to upgrade the dependencies used for testing https://github.com/wazuh/wazuh-jenkins/issues/5039
- Together with @mhamra we discussed and made the changes to solve the warnings found in the code review, they weren't detected before because of differences in the dependencies used in development and during the review

Changed the state to _pending review_.

**Comment by GGP1** [comment]
## Update

Rebased the branch to point to the epic [feature/13265-upgrade-python-to-3.10.11](https://github.com/wazuh/wazuh/tree/feature/13265-upgrade-python-to-3.10.11). The epic has failed checks so it's expected that the checks will fail in this one as well.

**Comment by GGP1** [comment]
## Update

Removed `py` test dependency.

The parent branch [feature/13265-upgrade-python-to-3.10.11](https://github.com/wazuh/wazuh/tree/feature/13265-upgrade-python-to-3.10.11) has modifications in the `Makefile` that cause some checks to fail, given that the Python 3.10.11 interpreter is already available, it's expected that this will be fixed soon. Moving the state to blocked until all checks pass.

**Comment by fdalmaup** [comment]
# Update

Since this issue involves changes in the `framework/requirements.txt` file, it is needed to update the modified dependencies of the Wazuh embedded Python interpreter to make these changes effective in Wazuh installations and to pass the checks of the associated PR. 
The update has been already carried out for the `amd64` architecture, the one for the `aarch64` architecture is in progress.

**Comment by fdalmaup** [comment]
## Update

The update of the Wazuh embedded Python interpreter dependencies has been done for the `aarch64` architecture and a clean installation of the Wazuh manager has been successfully carried out in CentOS 7.
The new files of the interpreter need to be uploaded in order for the [mentioned checks](https://github.com/wazuh/wazuh/issues/13267#issuecomment-1523820770) to pass.

**Comment by fdalmaup** [comment]
## Update

When installing Wazuh on Ubuntu 18.04, there was a failure due to not linking some of the Wazuh-developed libraries to the compilation of `hashlib` in `cpython/Modules/Setup` (new to `v3.10.11`). This led to regenerating the precompiled CPython files after making the corresponding modifications.
Once that was fixed, the `2.4+ kernel w/o ELF notes? -- report this` warning already seen in https://github.com/wazuh/wazuh/issues/16128#issuecomment-1462046167 was presented because some dependencies had been added prior to the findings of said issue. This problem will be addressed in #13265 because the involved dependencies do not correspond to the ones that arose from the unit test warnings.

The issue remains blocked until the fixing mentioned above and the update is carried out.

**Comment by mhamra** [comment]
### UPDATE
the rebase with the upstream branch was made. Failing checks are pending for the merge of PR [jenkins #5040](https://github.com/wazuh/wazuh-jenkins/pull/5040)

Reviewing the unit test results, three warnings were found and one exception. The warnings are related to the version of Python and the packages used by the pipeline, and it is expected to be solved once the mentioned PR is merged.

The failure was reproduced in a local environment and fixed. The root cause of the failure was the introduction of the cryptography package version 39.0.2 [see issue #16363](https://github.com/wazuh/wazuh/issues/16363). In this version, the function Fernet.encrypt doesn't call the function struct.pack like in the prior package version.

The following unit test mocks struct.pack and calls Handler.msg_build function. Inside Handler.msg_build function for the asserts 1, 2, and 4, Fernet.encrypt and struct.pack functions are called. The call count of the mocked struct.pack was 13 and now must be 10.

assert num|prior version|new version|
---|---|---|
Assert 1| 2 | 1
Assert 2| 2 | 1
Assert 3| 7 | 7
Assert 4| 2 | 1
**Total**|**13**|**10**

Updating the call_count to 10 fixes the failure on the unit test.

```
@patch('struct.pack', return_value=b"v1")
def test_handler_msg_build_ok(pack_mock):
    """Test if a message is being built with the right header and payload."""
    handler = cluster_common.Handler(fernet_key, cluster_items)

    # Test first if
    assert isinstance(handler.msg_build(b"command", 12345, b"data"), list)
    assert isinstance(handler.msg_build(b"command", 12345, b"data")[0], bytearray)

    # Test first else
    handler.header_len = 1
    handler.request_chunk = 20

    assert isinstance(handler.msg_build(b"command", 12345, b"000000000000000000000"), list)
    assert isinstance(handler.msg_build(b"command", 12345, b"data")[0], bytearray)

    assert pack_mock.call_count == 10

```
